pgjdbc/src/test/java/org/postgresql/jdbc/ResourceLockTest.java

```diff
@@ -0,0 +1,139 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.LockSupport;
+
+class ResourceLockTest {
+
+  @Test
+  void testObtainClose() {
+    final ResourceLock lock = new ResourceLock();
+
+    assertFalse(lock.isLocked(),
+        "lock.isLocked(). The newly created resource lock should be unlocked");
+    assertFalse(lock.isHeldByCurrentThread(),
+        "lock.isHeldByCurrentThread(). The newly created resource lock should not be held by the current thread");
+
+    try (ResourceLock ignore = lock.obtain()) {
+      assertTrue(lock.isLocked(),
+          "lock.isLocked(). Obtained lock should be locked");
+      assertTrue(lock.isHeldByCurrentThread(),
+          "lock.isHeldByCurrentThread(). Obtained lock should be held by the current thread");
+    }
+
+    assertFalse(lock.isLocked(), "lock.isLocked(). Closed resource lock should be unlocked");
+    assertFalse(lock.isHeldByCurrentThread(),
+        "lock.isHeldByCurrentThread(). Closed resource lock should not be held by the current thread");
+  }
+
+  @Test
+  void testObtainWhenMultiThreadedExpectLinearExecution() throws InterruptedException, ExecutionException {
+    CallWithResourceLock callWithResourceLock = new CallWithResourceLock();
+
+    int levelOfConcurrency = 5;
+
+    ExecutorService executorService = Executors.newFixedThreadPool(levelOfConcurrency);
+    try {
+      List<Callable<CounterPair>> callables = new ArrayList<>();
+      for (int i = 0; i < levelOfConcurrency; i++) {
+        callables.add(callWithResourceLock);
+      }
+
+      // expect linear execution
+      List<Future<CounterPair>> results = executorService.invokeAll(callables);
+
+      List<CounterPair> pairs = new ArrayList<>();
+      Set<Integer> preLockSet = new HashSet<>();
+      Set<Integer> postLockSet = new HashSet<>();
+      for (Future<CounterPair> result : results) {
+        CounterPair entry = result.get();
+        preLockSet.add(entry.preLock);
+        postLockSet.add(entry.postLock);
+        pairs.add(entry);
+      }
+
+      // confirm the callables were executed serially without any overlap
+      Collections.sort(pairs);
+      for (int i = 0; i < pairs.size() - 1; i++) {
+        CounterPair pair = pairs.get(i);
+        CounterPair nextPair = pairs.get(i + 1);
+        assertTrue(pair.beforeWithoutOverlap(nextPair), "executed serially without overlap");
+      }
```

Comment:
This has two flaws:
1) The first failure would hide the rest. In other words, you can't see if all of them overlap or not
2) The failure message includes no information on what really went wrong: any failure would look like `expected true, got false executed serially without overlap`. I am 100% sure the one who observes the failure would have absolutely no idea why something should be serial here, what went wrong, and so on.

Even though sometimes it requires to "reverse-engineer the test code in order to figure out the intention of the author", I am inclined the case for ResourceLock should not be like that. It should be possible to write a test that reports the failure in a human-comprehendable way that does not require reverse-engineering the test code.

One of the ways is to add full resulting collection to the failure message and clarify what do we expect from it.

Reply:
> The first failure would hide the rest. 

If this fails AT ALL / EVER we've hit a jdk bug.

Reply:
> If this fails AT ALL / EVER we've hit a jdk bug

I would prefer that the test code that is added to pgjdbc fails reasonably no matter if the cause is JDK bug or not.

Just in case, one of the possibilities for the failure is improper handling of the `nanoTime` overflow, which is not really a JDK bug.

Reply:
> fails reasonably no matter if the cause is JDK bug or not.

This is what it does currently. 

> improper handling of the nanoTime overflow

I don't see any problem here. You are obviously welcome to adjust or rewrite the test as you see fit.